core/txpool/validation.go

```diff
@@ -251,30 +252,35 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
 	}
 	// Ensure the transactor has enough funds to cover the transaction costs
 	var (
-		balance = opts.State.GetBalance(from).ToBig()
-		cost    = tx.Cost()
+		sgtBalanceSlot  = core.TargetSGTBalanceSlot(from)
+		sgtBalanceValue = opts.State.GetState(types.SoulGasTokenAddr, sgtBalanceSlot)
+		balance         = opts.State.GetBalance(from).ToBig()
+		cost            = tx.Cost()
 	)
+
+	sgtBalance := new(uint256.Int).SetBytes(sgtBalanceValue[:]).ToBig()
+
 	if opts.L1CostFn != nil {
 		if l1Cost := opts.L1CostFn(tx.RollupCostData()); l1Cost != nil { // add rollup cost
 			cost = cost.Add(cost, l1Cost)
 		}
 	}
-	if balance.Cmp(cost) < 0 {
-		return fmt.Errorf("%w: balance %v, tx cost %v, overshot %v", core.ErrInsufficientFunds, balance, cost, new(big.Int).Sub(cost, balance))
+	if balance.Cmp(cost) < 0 && sgtBalance.Cmp(cost) < 0 {
```

Comment:
The validation logic is not consistent with execution logic where msg.from is 0x0 (https://github.com/ethstorage/op-geth/blob/25669f6dbcda42fa8809094e53f164cc6369748c/core/state_transition.go#L342)

Reply:
Good catch, it seems the check for zero account is no longer necessary after we avoided evm.call, so removed. 

The UseSoulGasToken flag is added back.